docs/snapshots.md

```diff
@@ -0,0 +1,177 @@
+and a snapshotting technique based on the snapshot API Firecracker offers.
+
+There are 2 modes of snapshot operation: local and remote. While the local mode is stable and fully operational 
+(with a small issue, namely, GH-818), the remote mode is currently too [unstable](./snapshots.md#container-disk-state-restoration-blocker)
+to be used. 
+
+## Local snapshots
+
+The current approach allows loading an arbitrary amount of VMs from a single snapshot. This is done by creating a new
+shim and the required block and network devices upon loading a snapshot and creating an extra patch file containing the
+filesystem differences written by the VM upon snapshot creation. In addition, the snapshotting operation can be
+further configured using the following flags:
+
+- `netPoolSize [capacity]`: the amount of network devices in the Firecracker VM network pool (`10` by default), which
+  can be used to keep the network initialization off the cold start path of Firecracker VMs.
+
+### Snapshot creation
+
+Snapshots are created using the following algorithm.
+
+1. Pause the VM.
+2. Create a snapshot of the VM.
+3. Capture container snapshot (disk state) changes.
+   1. Get a snapshot of the original container image.
+   2. Mount the original container image snapshot.
+   3. Mount the current container snapshot.
+   4. Extract changes between the mounted container snapshots into a patch file using the `--only-write-batch` of `rsync`.
+4. Resume the VM.
+
+### Snapshot loading
+
+Snapshots are loaded using the following algorithm.
+
+1. Restore container snapshot (disk state) changes.
+   1. Get a snapshot of the original container image.
+   2. Mount the original container image snapshot.
+   3. Apply changes from the patch file to the mounter container snapshot using the `--read-batch` of `rsync`.
+2. Create a VM with a snapshot, providing the memory file, VM snapshot file and the path to the patched container snapshot. 
+
+## Remote snapshots
+
+Rather than only using the snapshots available locally on a node, snapshots can also be transferred between nodes to
+potentially accelerate cold start times and reduce memory utilization, given that proper mechanisms are in place to
+minimize the snapshot network transfer latency. This could be done by storing snapshots in a global storage solution
+such as [MinIO S3](./developers_guide.md#MinIO-S3-service), or directly distributing snapshots between compute nodes.
+
+### Snapshot creation
+
+Snapshots are created using the same algorithm as for local snapshots with an additional upload step (uploading 
+the snapshot files to the global storage solution).
+
+### Snapshot loading
+
+Snapshots are created using the same algorithm as for local snapshots with a preliminary download step (downloading
+the snapshot files from the global storage solution).
+
+### Blockers
+
+#### Container disk state restoration blocker
```

Comment:
can you please provide pointers to instructions on how to reproduce? we need a starting point for future contributors to pick up this task

Reply:
Fixed, created a [branch](https://github.com/vhive-serverless/vHive/tree/remote-firecracker-snapshots-poc) with the remote snapshots PoC, provided a [README](https://github.com/vhive-serverless/vHive/blob/remote-firecracker-snapshots-poc/remote-firecracker-snapshots-poc/README.md) with instructions on setting it up and working with it, added a [`Current state`](https://github.com/CuriousGeorgiy/vHive/blob/vanilla-firecracker-snapshots-impl/docs/snapshots.md#current-state) section to the snapshots document referring to this branch.